apps/src/lib/levelbuilder/lesson-editor/ActivitiesEditor.jsx

```diff
@@ -0,0 +1,121 @@
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+import color from '@cdo/apps/util/color';
+import ActivityCard from '@cdo/apps/lib/levelbuilder/lesson-editor/ActivityCard';
+import Activity from '@cdo/apps/templates/lessonOverview/activities/Activity';
+import {connect} from 'react-redux';
+import {addActivity} from '@cdo/apps/lib/levelbuilder/lesson-editor/activitiesEditorRedux';
+
+const styles = {
+  activityEditAndPreview: {
+    display: 'flex',
+    flexDirection: 'row'
+  },
+  editor: {
+    width: '50%'
+  },
+  preview: {
+    width: '45%',
+    marginLeft: 20
+  },
+  previewBox: {
+    border: '1px solid black',
+    padding: 10
+  },
+  addActivity: {
+    fontSize: 14,
+    color: 'white',
+    background: color.cyan,
+    border: `1px solid ${color.cyan}`,
+    boxShadow: 'none'
+  }
+};
+
+class ActivitiesEditor extends Component {
+  static propTypes = {
+    //redux
+    activities: PropTypes.array,
+    addActivity: PropTypes.func
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      targetActivitySectionPos: null
+    };
+  }
+
+  handleAddActivity = () => {
+    //To do - figure out how we want to set the key going forward
+    this.props.addActivity(
+      this.props.activities.length,
+      `activity-${this.props.activities.length + 1}`
```

Comment:
this seems like it might fail if a create activities `activity-1` and `activity-2`, delete `activity-1`, and then create a new activity. The user would be pretty stuck in this case if they had any content they cared about in other Activities, so I would consider some solution here to be blocking for January. I assume this problem is what your comment refers to but please correct me if I'm wrong. ideas for fixing:

1. keep trying higher numbers until you find one not matching an existing activity (probably easiest to start here)

2. prompt the user to type in an initial activity name like "Dave's Paper Planes", and then converting it into a key maybe like `dave-s-paper-planes`. Before we launch, we might have to layer technique (1) on top of this to ensure no duplicates.

3. forego an activity key and just number them sequentially. this may require some discussion, but my thinking is that we never need to reference them by a special key because (a) when seeded, these will live inside the .script_lessons file and (b) when translated, their strings will probably also be nested under a script object or possibly a lesson object. @uponthesun can you chime in on whether this analysis sounds right to you?

Reply:
Yes the comment was referring to the fact that this would not be the long term solution. I meant to follow up and ask about keys in general and if we wanted to start generating them for curriculum writers behind the scenes so they don't have to work about it. I'm open to any of the ideas you had mentioned here.

Reply:
I think I'm convinced that we don't need Activity or ActivitySection to have a `key` property in the database (@uponthesun , this is the part I'd like a sanity check on). Then it looks like the only thing you are using activity.key and section.key for is the React `key` property which is needed when you create elements in a loop like you are doing with `map`. you could try using the loop index (provided by `map`) as the react key and see if react complains. if it does complain, then you could instead choose any approach for generating a unique id for each element in the array of activities or activity sections, such as lodash's `_.uniqueId()`. you would need to set this in the UI only, whenever you load data from the server or like when you call addActivity here.

Reply:
The reason we might need a `key` is for seeding purposes, right? I think we don't need one if either:

* We don't need to identify and update existing `Activity` objects in the DB - all the relevant data is contained in the seeding file
* We have some other way of uniquely identifying it

Which of these is true in this case?

Reply:
Also, have we decided how we're going to model `Activity` and `ActivitySection` in the DB overall yet? I've seen the descriptions of the fields in the "Lesson Edit" page spec, but the data model is TBD still right? (We already have a model called `Activity` but I think it's for tracking user progress or something.)

Reply:
This is the one I think may be true:

> We don't need to identify and update existing Activity objects in the DB, like we do for other things that we seed

if we cannot find-and-update then we will destroy-and-recreate. because no other objects in the DB should need to have a reference directly to an Activity or ActivitySection, this shouldn't cause any data loss. to make sure seeding performance is not too slow, we could do a best-effort to find-and-update if the Activity or ActivitySection's name and array index are unchanged (this is a similar heuristic to how we seed script levels).

one could argue that this kind of optimization and reasoning adds more complexity to the system than just maintaining a unique key in the DB.

Thoughts on which seems better or worse?

Either option makes me sad that our seed process means we can't use the database id column, since it is not consistent across environments. but that's part of what we live with as long as we delay the LB Consolidation work.

> Also, have we decided how we're going to model Activity and ActivitySection in the DB overall yet? I've seen the descriptions of the fields in the "Lesson Edit" page spec, but the data model is TBD still right? 

correct, that is TBD. We may get clarify on some of this once we do that work.

> (We already have a model called Activity but I think it's for tracking user progress or something.)

yes, the Activity model today is something different. perhaps we could call the new model LessonActivity.

Reply:
> because no other objects in the DB should need to have a reference directly to an Activity or ActivitySection

If this is true, what do you think about just storing these as JSON objects in a property, rather than as actual DB models? I think we do this already for `script_announcements`. Do we consider that a good pattern?

Reply:
I think moving this to an in-person conversation might be easier. What do you both think?

Reply:
Sure, either is fine with me.

Reply:
>> because no other objects in the DB should need to have a reference directly to an Activity or ActivitySection

> what do you think about just storing these as JSON objects in a property, rather than as actual DB models?

I like this! Here is my best attempt to poke holes in this plan:

Today, Lesson `has_many :levels, through: :script_levels`. This provides the link between a level and which lessons it appears in. In the world where activities are part of a Lesson's properties, if we ever needed to find out which activity or progression (a type of activity section) a lesson was in, we'd have to use the `script_levels` link to find out which lessons it is in, and then search through its activities to find where exactly it appears in the lesson. This seems ok to me, especially because we are steering toward CourseVersion, Lesson and Level being the primary objects in the system. 

If we want to make it easier to access Activities in an object-oriented manner, we could construct LessonActivity objects from Lesson properties, kind of how we do for LevelGroupPage: https://github.com/code-dot-org/code-dot-org/blob/79ff09e81f6dfb2267d77b06106327665bc05fb6/dashboard/app/models/levels/level_group.rb#L63 that would let us write ruby expressions like `lesson.activities.first.sections.first` rather than having to parse the activities JSON each time.

How does this sound?

Reply:
> I think moving this to an in-person conversation might be easier. What do you both think?

Sorry, Dani, for mixing this discussion into your PR. Here is my request for this PR:

> it looks like the only thing you are using activity.key and section.key for is the React key property which is needed when you create elements in a loop like you are doing with map. you could try using the loop index (provided by map) as the react key and see if react complains. if it does complain, then you could instead choose any approach for generating a unique id for each element in the array of activities or activity sections, such as lodash's _.uniqueId(). you would need to set this in the UI only, whenever you load data from the server or like when you call addActivity here.

this shouldn't be hard to undo If we decide later that we need a unique key property on activity and activity section.

Reply:
@davidsbailey I have fixed all the keys to generate a key that is not being using for right now.